frame/nfts/src/features/attributes.rs

```diff
@@ -34,90 +35,273 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		// for the `CollectionOwner` namespace we need to check if the collection/item is not locked
+		match namespace {
+			AttributeNamespace::CollectionOwner => match maybe_item {
+				None => {
+					ensure!(
+						collection_config.is_setting_enabled(CollectionSetting::UnlockedAttributes),
+						Error::<T, I>::LockedCollectionAttributes
+					)
+				},
+				Some(item) => {
+					let maybe_is_locked = Self::get_item_config(&collection, &item)
+						.map(|c| c.has_disabled_setting(ItemSetting::UnlockedAttributes))?;
+					ensure!(!maybe_is_locked, Error::<T, I>::LockedItemAttributes);
+				},
 			},
-		};
+			_ => (),
+		}
 
-		let attribute = Attribute::<T, I>::get((collection, maybe_item, &key));
+		let attribute = Attribute::<T, I>::get((collection, maybe_item, &namespace, &key));
 		if attribute.is_none() {
 			collection_details.attributes.saturating_inc();
 		}
-		let old_deposit = attribute.map_or(Zero::zero(), |m| m.1);
-		collection_details.total_deposit.saturating_reduce(old_deposit);
+
+		let old_deposit =
+			attribute.map_or(AttributeDeposit { account: None, amount: Zero::zero() }, |m| m.1);
+
 		let mut deposit = Zero::zero();
-		if collection_config.is_setting_enabled(CollectionSetting::DepositRequired) &&
-			maybe_check_owner.is_some()
-		{
+		if collection_config.is_setting_enabled(CollectionSetting::DepositRequired) {
 			deposit = T::DepositPerByte::get()
 				.saturating_mul(((key.len() + value.len()) as u32).into())
 				.saturating_add(T::AttributeDepositBase::get());
 		}
-		collection_details.total_deposit.saturating_accrue(deposit);
-		if deposit > old_deposit {
-			T::Currency::reserve(&collection_details.owner, deposit - old_deposit)?;
-		} else if deposit < old_deposit {
-			T::Currency::unreserve(&collection_details.owner, old_deposit - deposit);
+
+		// NOTE: when we transfer an item, we don't move attributes in the ItemOwner namespace.
+		// When the new owner updates the same attribute, we will update the depositor record
+		// and return the deposit to the previous owner.
+		if old_deposit.account.is_some() && old_deposit.account != Some(origin.clone()) {
+			T::Currency::unreserve(&old_deposit.account.unwrap(), old_deposit.amount);
+			T::Currency::reserve(&origin, deposit)?;
+		} else if deposit > old_deposit.amount {
+			T::Currency::reserve(&origin, deposit - old_deposit.amount)?;
+		} else if deposit < old_deposit.amount {
+			T::Currency::unreserve(&origin, old_deposit.amount - deposit);
 		}
 
-		Attribute::<T, I>::insert((&collection, maybe_item, &key), (&value, deposit));
+		// NOTE: we don't track the depositor in the CollectionOwner namespace as it's always a
+		// collection's owner. This simplifies the collection's transfer to another owner.
+		let deposit_owner = match namespace {
+			AttributeNamespace::CollectionOwner => {
+				collection_details.owner_deposit.saturating_reduce(old_deposit.amount);
+				collection_details.owner_deposit.saturating_accrue(deposit);
```

Comment:
wondering if these two lines would be slightly safer switched around so that the accrue happened first?

Reply:
changed